### PR TITLE
[RFC] zynqmp: Add static TA for bitstream loading

### PR DESCRIPTION
**Pull request by sorenb-xlnx**
I created this patch for a proof of concept for loading a bitstream into the FPGA on ZynqMP devices through an OP-TEE TA (the HW for programming a bitstream allows secure accessses only).
Currently I'm just looking for general feedback on the overall solution. Though, I'd be inclined to go for a merge if feedback is positive.
The user-space portion is available in https://github.com/sorenb-xlnx/optee_bsload.

The high level flow is:
- userspace opens a bitstream (using mmap) and provides address and size as temp memref to OP-TEE (a second argument is used to pass flags, but so far none are implemented. But in the future cases like chunked or partial configuration may require some flags)
- the TA programs the HW to load the bitstream into the FPGA

### CODE REVIEW

File: core/arch/arm/plat-zynqmp/platform_config.h [hunk 1]
```diff
@@ -117,6 +119,18 @@
 #define DEVICE2_SIZE		CORE_MMU_DEVICE_SIZE
 #define DEVICE2_TYPE		MEM_AREA_IO_SEC
 
+#ifdef CFG_ZYNQMP_BSLOAD_STA
+# define DEVICE3_PA_BASE	ROUNDDOWN(CSUDMA_BASE, CORE_MMU_DEVICE_SIZE)
```
Comment (jenswi-linaro) [inline core/arch/arm/plat-zynqmp/platform_config.h]: Please use `register_phys_mem()` instead, for instance in `core/arch/arm/plat-zynqmp/sta_bsload.c`.
Same for `DEVICE4_*` below.

File: core/arch/arm/plat-zynqmp/sta_bsload.c [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+		reg = read32(addr);
+	} while (((reg & mask) != val) && tout--);
+
+	if (tout <= 0)
+		return TEE_ERROR_GENERIC;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result check_header(const paddr_t buf)
```
Comment (jenswi-linaro) [inline core/arch/arm/plat-zynqmp/sta_bsload.c]: s/paddr_t/vaddr_t/

File: core/arch/arm/plat-zynqmp/sta_bsload.c [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+		return ret;
+
+	addr = virt_to_phys((void *)bs);
+	if (!addr) {
+		DMSG("no mapping found for va '%#lx'\n", bs);
+		return TEE_ERROR_GENERIC;
+	}
+	DMSG("bitstream va:%#lx, pa:%#lx, sz:%#zx\n", bs, addr, sz);
+
+	flush_dcache_range(bs, sz);
```
Comment (jenswi-linaro) [inline core/arch/arm/plat-zynqmp/sta_bsload.c]: Please use `cache_maintenance_l1(DCACHE_AREA_CLEAN_INV, (void *)bs, sz)` instead.

File: core/arch/arm/plat-zynqmp/sta_bsload.c [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+{
+	uint32_t exp_param_types = TEE_PARAM_TYPES(TEE_PARAM_TYPE_NONE,
+						   TEE_PARAM_TYPE_NONE,
+						   TEE_PARAM_TYPE_NONE,
+						   TEE_PARAM_TYPE_NONE);
+	if (param_types != exp_param_types)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	/* Unused parameters */
+	(void)&params;
```
Comment (jenswi-linaro) [inline core/arch/arm/plat-zynqmp/sta_bsload.c]: Use the `__unused` attribute in the function header instead.

File: core/arch/arm/plat-zynqmp/sta_bsload.c [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+						   TEE_PARAM_TYPE_VALUE_INPUT,
+						   TEE_PARAM_TYPE_NONE,
+						   TEE_PARAM_TYPE_NONE);
+
+	if (param_types != exp_param_types)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	DMSG("bitstream va:%p, size:%#x\n",
+			params[0].memref.buffer, params[0].memref.size);
+	TEE_ASSERT(params[0].memref.buffer && params[0].memref.size);
```
Comment (jenswi-linaro) [inline core/arch/arm/plat-zynqmp/sta_bsload.c]: Better to return an error here than panic the system.

File: core/arch/arm/plat-zynqmp/sta_bsload.c [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+
+static TEE_Result pcap_write(uint64_t srcaddr, uint32_t size)
+{
+	TEE_Result ret;
+	uint32_t regval;
+
+	write32(0, CSUDMA_SRC_CRC);
+
+	regval = read32(CSU_SSS_CONFIG);
+	regval &= CSU_SSS_CFG_PCAP_SSS_MASK;
```
Comment (jenswi-linaro) [inline core/arch/arm/plat-zynqmp/sta_bsload.c]: I suppose you meant `~(CSU_SSS_CFG_PCAP_SSS_MASK << CSU_SSS_CFG_PCAP_SSS_SHFT)`.
Consider using the helper function `io_mask32()` instead.

File: core/arch/arm/plat-zynqmp/sta_bsload.c [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+	regval |= (CSU_SSS_CFG_PCAP_DMA_DATA << CSU_SSS_CFG_PCAP_SSS_SHFT);
+	write32(regval, CSU_SSS_CONFIG);
+
+	write32((uint32_t)srcaddr, CSUDMA_SRC_ADDR);
+	write32((srcaddr >> 32) & 0x1FFFF, CSUDMA_SRC_ADDR_MSB);
```
Comment (jenswi-linaro) [inline core/arch/arm/plat-zynqmp/sta_bsload.c]: `0x1FFFF` looks like a magic number.

File: core/arch/arm/plat-zynqmp/sta_bsload.c [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+	write32(size & CSUDMA_SRC_SIZE_MASK, CSUDMA_SRC_SIZE);
+
+	ret = poll_reg_tout_32(CSUDMA_SRC_I_STS, CSUDMA_SRC_I_STS_DONE,
+			       CSUDMA_SRC_I_STS_DONE, 0x8000000);
```
Comment (jenswi-linaro) [inline core/arch/arm/plat-zynqmp/sta_bsload.c]: `0x8000000` looks like a magic number

File: core/arch/arm/plat-zynqmp/sta_bsload.c [hunk 9]
```diff
@@ -0,0 +1,366 @@
...
+
+	write32(CSU_PCAP_CTRL_PCAP_PR, CSU_PCAP_CTRL);
+	write32(0, CSU_PCAP_RDWR);
+
+	/* reset PL */
+	write32(0, CSU_PCAP_PROG);
+	ret = poll_reg_tout_32(CSU_PCAP_STATUS, CSU_PCAP_STATUS_PL_RESET_B, 0,
+			       0x8000000);
+	if (ret != TEE_SUCCESS) {
+		DMSG("timeout waiting for PL reset\n");
```
Comment (jenswi-linaro) [inline core/arch/arm/plat-zynqmp/sta_bsload.c]: Strings for `DMSG()` and friends doesn't need a `\n` at the end.

File: core/arch/arm/plat-zynqmp/sta_bsload.c [hunk 10]
```diff
@@ -0,0 +1,366 @@
...
+}
+
+static TEE_Result do_load_bs(vaddr_t bs, size_t sz, uint64_t flags)
+{
+	paddr_t addr;
+	TEE_Result ret;
+	(void)flags;
+
+	/* sanity check bitstream */
+	if (bs & 0x3)
```
Comment (jenswi-linaro) [inline core/arch/arm/plat-zynqmp/sta_bsload.c]: Please use `!ALIGNMENT_IS_OK(bs, uint32_t)` instead.

File: core/arch/arm/plat-zynqmp/sta_bsload.c [hunk 11]
```diff
@@ -0,0 +1,366 @@
...
+	if (sz > BS_SIZE_MAX)
+		return TEE_ERROR_EXCESS_DATA;
+
+	ret = check_header(bs);
+	if (ret)
+		return ret;
+
+	addr = virt_to_phys((void *)bs);
+	if (!addr) {
+		DMSG("no mapping found for va '%#lx'\n", bs);
```
Comment (jenswi-linaro) [inline core/arch/arm/plat-zynqmp/sta_bsload.c]: Please use `PRIxVA` instead of `"lx"` for `vaddr_t` types.

**Comment by jenswi-linaro** [comment]
Overall I think this is the way to do stuff like this.

File: core/arch/arm/plat-zynqmp/sta_bsload.c (hunk 3 above)
Comment (sorenb-xlnx) [inline core/arch/arm/plat-zynqmp/sta_bsload.c]: I'll look into this, but L1 cleaning is not enough. The data needs to be all the way out on DRAM for the DMA to be able to pick it up. Judging by the function name, I have some doubts this change will do that.

Comment (jenswi-linaro) [inline core/arch/arm/plat-zynqmp/sta_bsload.c]: The name of the function is misleading, it handles the internal caches. `flush_dcache_range()` is only available in Aarch64.

Comment (sorenb-xlnx) [inline core/arch/arm/plat-zynqmp/sta_bsload.c]: OK, thanks for explaining. I'll switch.

File: core/arch/arm/plat-zynqmp/sta_bsload.c (hunk 6 above)
Comment (sorenb-xlnx) [inline core/arch/arm/plat-zynqmp/sta_bsload.c]: On Wed, 2016-07-20 at 00:26:38 -0700, Jens Wiklander wrote:
> > +		DMSG("timeout waiting for PL init done\n");
> > +		return ret;
> > +	}
> > +
> > +	return TEE_SUCCESS;
> > +}
> > +
> > +static TEE_Result pcap_write(uint64_t srcaddr, uint32_t size)
> > +{
> > +	TEE_Result ret;
> > +	uint32_t regval;
> > +
> > +	write32(0, CSUDMA_SRC_CRC);
> > +
> > +	regval = read32(CSU_SSS_CONFIG);
> > +	regval &= CSU_SSS_CFG_PCAP_SSS_MASK;
> 
> I suppose you meant `~(CSU_SSS_CFG_PCAP_SSS_MASK << CSU_SSS_CFG_PCAP_SSS_SHFT)`.
> Consider using the helper function `io_mask32()` instead.

The mask is already shifted, but the negation was missing. I'll switch
to the io_mask32 function where possible.

**Comment by sorenb-xlnx** [comment]
Thanks @jenswi-linaro for reviewing. I fixed up all the issues you pointed out. And it's really good to know that this approach is in line with what you think.
I'll close the PR for now. But I may re-open it as I have the feeling this may actually be a valuable feature.

File: core/arch/arm/plat-zynqmp/sta_bsload.c (hunk 6 above)
Comment (vchong) [inline core/arch/arm/plat-zynqmp/sta_bsload.c]: Just a quick note that io_mask* functions expect that val and mask to be already shifted as well.

**Pull request closed without merging**